Master adxl372 support

stefpopa opened the pull request:
This driver was initially implemented on the [rpi-4.9.y](https://github.com/analogdevicesinc/linux/blob/rpi-4.9.y/drivers/iio/accel/adxl372.c) branch. In order to add it to the master branch a few changes were made:
* fixed the commit history 
* deleted unused macros
* fixed sparse warnings
* various style fixes

Adding the driver to the master branch is the preceding step to sending it upstream.

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -207,14 +207,23 @@ static const int adxl372_samp_freq_tbl[5] = {
+	[ADXL372_BW_400HZ] = 400,
+	[ADXL372_BW_800HZ] = 800,
+	[ADXL372_BW_1600HZ] = 1600,
+	[ADXL372_BW_3200HZ] = 3200,
```
you could remove the enums here and just list the frequencies in order
it doesn't seem like the enum names are used

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -598,6 +619,11 @@ static int adxl372_read_raw(struct iio_dev *indio_dev,
+
+	case IIO_CHAN_INFO_LOW_PASS_FILTER_3DB_FREQUENCY:
+		*val = adxl372_bw_freq_tbl[st->bw];
+
```
nitpick: extra line

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -0,0 +1,493 @@
+#include <linux/iio/sysfs.h>
+
+/* ADXL372 registers definition */
+#define ADXL372_DEVID			0x00
```
a lot of these enums and definitions are used later in this commit series;
that makes things a bit harder to follow & evaluate;

typically for Linux drivers, when adding them you only add the minimum necessary [of definitions]

i believe some enums could go away, because they could be redundant; some enums have values 0,1,2,...,N and are used to index arrays

mhennerich replied:
A lot of maintainer don't mind having a full register defines.

commodo replied:
ack ;
i don't (yet) have a good feel for the preferences in the IIO subsystem

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -0,0 +1,493 @@
+	ADXL372_THRESHOLD_EN,
+};
+
+enum adxl372_threshold_ref_en {
```
this enum looks like could be replaced with a bool or int enable ( 0 = off, 1 = on )  type

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -0,0 +1,493 @@
+ */
+#define ADXL372_USCALE	958241
+
+enum adxl372_axis {
```
this looks unused here

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -171,13 +181,40 @@ enum adxl372_bandwidth {
+	ADXL372_FIFO_OLD_SAVED
+};
+
+static const int adxl372_samp_freq_tbl[5] = {
```
you could remove the enum and just list these values

commodo commented:
Let's add the driver to the Kconfigs `zynq_xcomm_adv7511_defconfig` & `adi_zynqmp_defconfig` ; these 2 are generic platforms in master that build ADI drivers/stuff

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -461,14 +833,48 @@ static int adxl372_probe(struct spi_device *spi)
+	if (ret < 0)
+		return ret;
+
+	iio_buffer_set_attrs(indio_dev->buffer, adxl372_fifo_attributes);
```
i think you may need to backport this patch into master as well; to support this function ;

commodo replied:
my bad; found int the series

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -418,14 +595,209 @@ static int adxl372_read_raw(struct iio_dev *indio_dev,
+		 * The maximum bandwidth is constrained to at most half of
+		 * the ODR to ensure that the Nyquist criteria is not violated
+		 */
+		if (st->bw > odr_index)
```
should we compare `st->bw > (st->odr / 2) ` here ?
it would be a bit more aligned with the comment;

stefpopa replied:
Here I am comparing indexes and not actual bw and odr values:

enum adxl372_odr {
	ADXL372_ODR_400HZ,
	ADXL372_ODR_800HZ,
	ADXL372_ODR_1600HZ,
	ADXL372_ODR_3200HZ,
	ADXL372_ODR_6400HZ,
};

enum adxl372_bandwidth {
	ADXL372_BW_200HZ,
	ADXL372_BW_400HZ,
	ADXL372_BW_800HZ,
	ADXL372_BW_1600HZ,
	ADXL372_BW_3200HZ,
};

commodo replied:
ah ; thanks for the clarification ; 
my bad for not noticing

mirceacaprioru commented on drivers/iio/accel/adxl372.c:
```diff
@@ -418,14 +595,209 @@ static int adxl372_read_raw(struct iio_dev *indio_dev,
+static IIO_CONST_ATTR(hwfifo_watermark_min, "1");
+static IIO_CONST_ATTR(hwfifo_watermark_max,
+		      __stringify(ADXL372_FIFO_SIZE));
+static IIO_DEVICE_ATTR(hwfifo_watermark, 0444,
```
could you replace 0444 with S_IRUGO. it is more transparent.

lclausen-adi replied:
I think upstream is going the opposite direction. Replacing all S_... with the octal number representation.

mirceacaprioru replied:
got it. Thanks.

stefpopa commented:
changes in v2:
* added the full register defines
* removed unused and unnecessary enums
* used a `adxl372_find_closest_match()` when looking for the odr and bw indexes
* added the driver to `zynq_xcomm_adv7511_defconfig `& `adi_zynqmp_defconfig`
* added the `iio:adxl372: Add sampling frequency support` commit which was accidentally squashed into the previous commit

commodo commented on Documentation/devicetree/bindings/iio/accel/adxl372.txt:
```diff
@@ -0,0 +1,16 @@
+Analog Devices ADXL372 3-Axis, +/-(200g) Digital Accelerometer
+
+http://www.analog.com/media/en/technical-documentation/data-sheets/adxl372.pdf
```
nitpick: commit requires description;

the documentation could also use a bit of wording

maybe some short adaptation from the datasheet would work in the documentation

commodo replied:
hmm ; looking at other dt bindings docs it looks like this is fine;
i was confusing this with other docs/types

stefpopa commented:
changes in v3:
* added a commit description for `dt-bindings: iio: accel: Add docs for ADXL372`  commit
